imjoy/helper.py

```diff
@@ -122,48 +122,49 @@ def parseEnv(eng, env, work_dir, default_env_name):
-                        is_py2 = True
-                    parms = shlex.split(_env)
-                    if "-n" in parms:
-                        env_name = parms[parms.index("-n") + 1]
-                    elif "--name" in parms:
-                        env_name = parms[parms.index("--name") + 1]
-                    else:
-                        env_name = default_env_name
-                        envs[i] = _env.replace(
-                            "conda create", "conda create -n " + env_name
-                        )
-
-                    if "-y" not in parms:
-                        envs[i] = _env.replace("conda create", "conda create -y")
-
-                if "conda env create" in _env:
-                    parms = shlex.split(_env)
-                    if "-f" in parms:
-                        try:
-                            env_file = os.path.join(
-                                work_dir, parms[parms.index("-f") + 1]
+                if type(_env) is str:
+                    if "conda create" in _env:
+                        # if not _env.startswith('conda'):
+                        #     raise Exception('env command must start with conda')
+                        if "python=2" in _env:
+                            is_py2 = True
+                        parms = shlex.split(_env)
+                        if "-n" in parms:
+                            virtual_env_name = parms[parms.index("-n") + 1]
+                        elif "--name" in parms:
+                            virtual_env_name = parms[parms.index("--name") + 1]
+                        else:
+                            virtual_env_name = default_env_name
+                            envs[i] = _env.replace(
+                                "conda create", "conda create -n " + virtual_env_name
                             )
-                            with open(env_file, "r") as stream:
-                                env_config = yaml.load(stream)
-                                assert "name" in env_config
-                                env_name = env_config["name"]
-                        except Exception as exc:
+
+                        if "-y" not in parms:
+                            envs[i] = _env.replace("conda create", "conda create -y")
+
+                    if "conda env create" in _env:
+                        parms = shlex.split(_env)
+                        if "-f" in parms:
+                            try:
+                                env_file = os.path.join(
+                                    work_dir, parms[parms.index("-f") + 1]
+                                )
+                                with open(env_file, "r") as stream:
+                                    env_config = yaml.load(stream)
+                                    assert "name" in env_config
+                                    virtual_env_name = env_config["name"]
+                            except Exception as exc:
```

Comment:
What errors are we expecting?

Reply:
Well, the file not exists, not a yaml file, yaml file corrupted, don't contain name in the config etc.

Reply:
We really need to get away from catching broad exceptions all the time.

Name in config, we should validate in the frontend, right?

For the remaining things, we should except just:
`FileNotFoundError`
`yaml.YAMLError`
`UnicodeDecodeError`

Excepting specific errors also makes it possible to have specific error messages.

Reply:
Not for this case, it's read from a file which only exist in the engine side.

 While I agree that it's better to not use broad exceptions all the time, but that would means we need to really predict every type of error, very strong tests covers all the possible cases. Obviously that's not possible to predict, especially when we have no test code at the moment. 

In many cases, I think we should more focus on the stability of the engine rather than code style, that means, even if we believe that's all the error types, we should at least add broad exception in the end. For error types we don't really care, why don't just use a single broad exception? 

One extreme case is a simple typo in the code (happened last night :( ), which we don't expect, and we deployed it. The GPUtil code fails and it blow up the entire engine, but what I actually trying to do is to get gpu information, which is not critical.

I think when we have more user, and also because the engine will always upgrade itself, we should really make sure it's stable.

Reply:
What do you think about ways for improving stability (besides tests which we are going to do)?

Reply:
It's better to let it blow up, we notice the problem and we can fix it. Broad exceptions hide the problems and mask them, making them harder to fix.

In our own code we can predict the common errors. Other errors will surface if we don't catch them and then we can fix them or handle them specifically.

I think we're working against ourselves by having this habit of catching broad exceptions.

It's not code style what errors you catch, in the sense it doesn't change the behavior of your code. This has real impact. If we want to get a stable engine we need to be able to see and fix problems.

Reply:
At this stage I think it's ok to let us and users find errors in releases, since we're small and have frequent releases. We can usually fix bugs and make a new release quickly.

At a later stage, we can eg start having a beta release and then a fter a period of time make the beta stable. The time period we can adjust. Betas will increase the burden of creating releases though.

Reply:
I think we are talking about slightly different thing here, we don't differentiate the error types doesn't necessary we hide the error, we do log them. We always log the error, and we will notice it in our test or user's bug report right?

For ourselves and experienced users with programming skills, I agree that we can do that, and we may get bug report etc. 

However, there is the other side of the problem, the engine meant to be used by unexperienced users, which you never want to crash, if that happens, they cannot use it and give up. We will then lose the user and without knowing it and no bug report. So we really don't want to crash.

Moreover, a bug in the engine itself will less likely to change the result of the analysis running on it. Most cases, without crashing the engine, we can easily notice it since many of the error logs are forwarded to the interface.